substrate/primitives/blockchain/src/backend.rs

```diff
@@ -226,88 +226,124 @@ pub trait Backend<Block: BlockT>:
 		finalized_block_hash: Block::Hash,
 		finalized_block_number: NumberFor<Block>,
 	) -> std::result::Result<DisplacedLeavesAfterFinalization<Block>, Error> {
-		let mut result = DisplacedLeavesAfterFinalization::default();
-
 		let leaves = self.leaves()?;
 
 		// If we have only one leaf there are no forks, and we can return early.
 		if finalized_block_number == Zero::zero() || leaves.len() == 1 {
-			return Ok(result)
+			return Ok(DisplacedLeavesAfterFinalization::default())
 		}
 
-		let first_leaf = leaves.first().ok_or(Error::Backend(
-			"Unable to find any leaves. This should not happen.".to_string(),
-		))?;
-		let leaf_block_header = self.expect_header(*first_leaf)?;
-
-		// If the distance between the leafs and the finalized block is large,  calculating
-		// tree routes can be very expensive. In that case, we will try to find the
-		// lowest common ancestor between all the leaves. The assumption here is that the forks are
-		// close to the tip and not long. So the LCA can be computed from the header cache. If the
-		// LCA is above the finalized block, we know that there are no displaced leaves by the
-		// finalization.
-		if leaf_block_header
-			.number()
-			.checked_sub(&finalized_block_number)
-			.unwrap_or(0u32.into()) >
-			header_metadata::LRU_CACHE_SIZE.into()
-		{
-			if let Some(lca) = lowest_common_ancestor_multiblock(self, leaves.clone())? {
-				if lca.number > finalized_block_number {
-					return Ok(result)
-				} else {
-					log::warn!("The distance between leafs and finalized block is large. Finalization can take a long time.");
-				}
-			};
-		}
+		// Store hashes of finalized blocks for quick checking later, the last block if the
+		// finalized one
+		let mut finalized_chain = VecDeque::new();
+		finalized_chain.push_front(self.header_metadata(finalized_block_hash)?);
+
+		// Local cache is a performance optimization in case of finalized block deep below the
+		// tip of the chain with a lot of leaves above finalized block
+		let mut local_cache = HashMap::<Block::Hash, CachedHeaderMetadata<Block>>::new();
```

Comment:
One thing I thought about:

This cache will become quite memory intensive when finalizing large gaps. For parachains which sync from scratch but use and external relay chain RPC, it is likely that they finalize from genesis to tip. If there happen to be multiple branches at finalization time, the `leaves.len() == 1` optimization above will not be taken.

So for each year that goes by, these chains need ~800MB more memory to keep this data in the cache here. This problem can be avoided by using warp sync, but still. We will have chains with 3 and 2 seconds block time which will grow even faster.

So what we could do:
1. Only cache what we need. Since we only need the hash, number and parent hash we can save around 40% here.
2. Use `schnellru` instead of the HashMap and set a byte limit (only peek to not use LRU capability), can be generous like 2GB or something. This would still be a lot faster than relying on the relatively small `HeaderMetadataCache`.
3. Abandon caching here totally. Initial finalization will take a while, but will be only called once instead of for every block. Operation at chain tip will still benefit from `HeaderMetadataCache`.
4. Accept that this requires an ever growing amount of memory.

I think 2 is good and easy to integrate, with 1 as optional optimization.